### PR TITLE
sol

### PR DESCRIPTION
**Pull request by NikitaNikolaitsev**

### CODE REVIEW

File: crud.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    return db.query(models.DBAuthor).filter(
+        models.DBAuthor.id == author_id
+    ).first()
+
+
+def create_new_author(db: Session, author: schema.AuthorCreate):
+    db_author = models.DBAuthor(
+        name=author.name,
+        bio=author.bio
+    )
```
Comment (DevSpaciX) [inline crud.py]: Provide return type annotations in all places

**Review by DevSpaciX — CHANGES_REQUESTED** [review]

**Review by DevSpaciX — CHANGES_REQUESTED** [review]
And remove db file from PR please )

File: crud.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    db_book = models.DBBook(
+        title=book.title,
+        description=book.description,
+        publication_date=book.publication_date,
+        author_id=book.author_id,
+    )
+    db.add(db_book)
+    db.commit()
+    db.refresh(db_book)
+    return db_book
```
Comment (Arsen-hrynevych) [inline crud.py]: Where is try block?

File: database.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import sessionmaker
+
+DATABASE_URL = "sqlite:///./library.db"
+
+engine = create_engine(
+    DATABASE_URL, connect_args={"check_same_thread": False}
+)
+
+SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)
```
Comment (Arsen-hrynevych) [inline database.py:11]: Blank line

File: main.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+from fastapi import FastAPI, Depends, HTTPException
+from sqlalchemy.orm import Session
+
+import schema
+import crud
+from database import SessionLocal
+
```
Comment (Arsen-hrynevych) [inline main.py:11]: 2 blank lines

File: main.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        db: Session = Depends(get_db)
+):
+    return crud.get_all_books(db, skip=skip, limit=limit)
+
+@app.get("/books/", response_model=list[schema.Book])
+def create_book(
+        book: schema.BookCreate,
+        db: Session = Depends(get_db)
+):
+    return crud.create_new_book(db, book=book)
```
Comment (Arsen-hrynevych) [inline main.py]: add return type hint

File: models.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+from sqlalchemy import Column, Integer, String, Date, ForeignKey
+from sqlalchemy.orm import relationship
+
+from database import Base
+
+
+class DBAuthor(Base):
```
Comment (Arsen-hrynevych) [inline models.py]: Rename it to `Author`. We don't use `DBAuthor` to indicate it's a model; in FastAPI, we differentiate using `model.Author`, `schemas.Author`, etc.

File: models.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    __tablename__ = "author"
+
+    id = Column(Integer, primary_key=True, index=True)
+    name = Column(String(255), nullable=False, unique=True)
+    bio = Column(String(511), nullable=False, unique=False)
+
+    books = relationship("DBBook", back_populates="author")
+
+
+class DBBook(Base):
```
Comment (Arsen-hrynevych) [inline models.py]: Rename it to `Book`. We don't use `DBBook` to indicate it's a model; in FastAPI, we differentiate using `model.Book`, `schemas.Book`, etc.

File: schema.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+class AuthorCreate(BaseAuthor):
+    pass
+
+
+class Author(BaseModel):
+    id: int
+
+    class Config:
+        orm_mode = True
```
Comment (Arsen-hrynevych) [inline schema.py]: It's deprecated: replace `orm_mode` -> `from_attributes`.

File: schema.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+class BookCreate(BaseBook):
+    pass
+
+
+class Book(BaseModel):
+    id: int
+
+
+    class Config:
+        orm_mode = True
```
Comment (Arsen-hrynevych) [inline schema.py]: It's deprecated: replace `orm_mode` -> `from_attributes`.

**Review by Arsen-hrynevych — CHANGES_REQUESTED** [review]
Thanks

File: alembic/env.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+# Interpret the config file for Python logging.
+# This line sets up loggers basically.
+if config.config_file_name is not None:
+    fileConfig(config.config_file_name)
+
+# add your model's MetaData object here
+# for 'autogenerate' support
+# from myapp import mymodel
+# target_metadata = mymodel.Base.metadata
+target_metadata = [Base.metadata]
```
Comment (maxgerman) [inline alembic/env.py]: doesn't need to be a list

File: crud.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
+from sqlalchemy.orm import Session
+from typing import List, Optional
+
+import models
+import schema
+
+def get_all_authors(db: Session) -> List[models.Author]:
+    return db.query(models.Author).all()
```
Comment (maxgerman) [inline crud.py]: `query` is outdated style - check out the current docs (>2.0)

File: crud.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
+from sqlalchemy.orm import Session
+from typing import List, Optional
+
+import models
+import schema
+
+def get_all_authors(db: Session) -> List[models.Author]:
```
Comment (maxgerman) [inline crud.py]: this needs the same pagination params as books, so these can be implemented as a fastapi dependency

File: main.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+def read_author(
+        author_id: int, db: Session = Depends(get_db)
+) -> schema.Author:
+    db_author = crud.get_author_by_id(db, author_id)
+    if db_author is None:
+        raise HTTPException(status_code=404, detail="Author not found")
+    return db_author
+
+@app.post("/authors/", response_model=schema.Author)
+def create_author(
```
Comment (maxgerman) [inline main.py:49]: don't forget the relevant http status codes

File: main.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        db: Session = Depends(get_db)
+) -> List[schema.Book]:
+    return crud.get_all_books(db, skip=skip, limit=limit)
+
+@app.post("/books/", response_model=schema.Book)
+def create_book(
+        book: schema.BookCreate,
+        db: Session = Depends(get_db)
+) -> schema.Book:
+    return crud.create_new_book(db, book=book)
```
Comment (maxgerman) [inline main.py]: what if the book already exists?

File: schema.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    bio: str
+
+class AuthorCreate(BaseAuthor):
+    pass
+
+
+class Author(BaseModel):
+    id: int
+
+    class Config:
```
Comment (maxgerman) [inline schema.py]: this is the old style (pre-2.0) - see the docs - there's a migration guide in pydantic 2.0 docs specifically for such things

**Review by maxgerman — CHANGES_REQUESTED** [review]
Good job!
The async style is recommended (session, db requests) - it's a perfect match for fastapi

File: crud.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+import schema
+
+def get_all_authors(db: Session, skip: int = 0, limit: int = 10) -> List[models.Author]:
+    stmt = select(models.Author).offset(skip).limit(limit)
+    result = db.execute(stmt)
+    return result.scalars().all()
+
+
+def get_author_by_id(db: Session, author_id: int) -> Optional[models.Author]:
+    return db.query(models.Author).filter(
```
Comment (maxgerman) [inline crud.py]: query is outdated style - check out the current docs (>2.0)

File: main.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        yield db
+    finally:
+        db.close()
+
+app = FastAPI()
+
+
+@app.get("/authors/", response_model=List[schema.Author])
+def read_authors(
+        limit: int = 10, skip: int = 0, db: Session = Depends(get_db)
```
Comment (maxgerman) [inline main.py]: the same pagination params are used in another endpoint - it's better to refactor them as a fastapi dependecy

File: main.py [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Book with this title already exists."
+        )
+
+    try:
+        new_book = crud.create_new_book(db, book=book)
+        return new_book
+    except IntegrityError:
+        db.rollback()
```
Comment (maxgerman) [inline main.py]: you don't have to do a `rollback` here, if creation fails the db will be intact. Also, you've already handled the main reason for IntegrityError here

**Review by maxgerman — CHANGES_REQUESTED** [review]
see comments and also see some previous comments still unhandled

File: crud.py [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+        bio=author.bio
+    )
+    db.add(db_author)
+    db.commit()
+    db.refresh(db_author)
+    return db_author
+
+
+def get_all_books(db: Session, skip: int = 0, limit: int = 10) -> List[models.Book]:
+    query = db.query(models.Book)
```
Comment (maxgerman) [inline crud.py]: same

**Review by maxgerman — CHANGES_REQUESTED** [review]

File: main.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+        raise HTTPException(status_code=409, detail="Author already exists")
+    return crud.create_new_author(db, author=author)
+
+
+@app.get("/books/", response_model=List[schema.Book])
+def read_books(
+        pagination: PaginationParams = Depends(),
+        db: Session = Depends(get_db)
+) -> List[schema.Book]:
+    return crud.get_all_books(db, skip=pagination.skip, limit=pagination.limit)
```
Comment (maxgerman) [inline main.py:64]: (optional) you could even pass the whole pagination obj to the crud function and let it handle it

File: main.py (hunk 13 above)
Comment (maxgerman) [reply main.py:49]: "created" status code?

File: main.py [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+    if existing_book:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail="Book with this title already exists."
+        )
+
+    try:
+        new_book = crud.create_new_book(db, book=book)
+        return new_book
+    except IntegrityError:
```
Comment (maxgerman) [inline main.py]: what are you trying to catch here?

File: main.py [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+        raise HTTPException(
+            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
```
Comment (maxgerman) [inline main.py]: usually we don't return 500 ourselves - if we expect something can go wrong we handle these scenarios and return the corresponding error messages and status codes

**Review by maxgerman — CHANGES_REQUESTED** [review]
old comments are not fixed + see some new

**Review by Arsen-hrynevych — APPROVED** [review]
Seems ok

**Pull request closed without merging**